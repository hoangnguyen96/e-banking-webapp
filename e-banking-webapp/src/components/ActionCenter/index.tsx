'use client';

import { Card, CardBody, CardHeader } from '@nextui-org/react';
import { useState } from 'react';

// Interfaces
import { TransferType } from '@/interfaces';

// Components
import { MetricsCard, Text, TransferTable } from '@/components';

export const ActionCenter = () => {
  const [selectedTransferType, setSelectedTransferType] = useState<
    TransferType.RECEIVED | TransferType.SENT
  >(TransferType.RECEIVED);

  const handleSelectReceived = () => {
    setSelectedTransferType(TransferType.RECEIVED);
  };

  const handleSelectSent = () => {
    setSelectedTransferType(TransferType.SENT);
  };

  return (
    <Card className='gap-6 py-5 pl-5 pr-[38px]'>
      <CardHeader className='flex-col items-start gap-6 p-0'>
        <Text as='span'>Action Center</Text>
        <div className='ml-[50px] flex gap-6'>
          <div className='cursor-pointer' onClick={handleSelectReceived}>
            <MetricsCard
              // TODO: Props totalTransfers, percentageChange will get data from API
              title='Transfer Received'
              totalTransfers={54}
              isPositive={true}
              percentageChange={8}
              isSelected={selectedTransferType === TransferType.RECEIVED}
            />
          </div>
          <div className='cursor-pointer' onClick={handleSelectSent}>
            <MetricsCard
              // TODO: Props totalTransfers, percentageChange will get data from API
              title='Transfer Sent'
              totalTransfers={14}
              isPositive={true}
              percentageChange={8}
              isSelected={selectedTransferType === TransferType.SENT}
            />
          </div>
        </div>
      </CardHeader>
      <CardBody>
        {/* TODO: Prop transactions will get data from API */}
        <TransferTable transactions={[]} transferType={selectedTransferType} />
      </CardBody>
    </Card>
  );
};
