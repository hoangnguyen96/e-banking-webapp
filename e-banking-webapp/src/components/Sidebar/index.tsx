'use client';

// Libs
import Link from 'next/link';
import { usePathname } from 'next/navigation';

// Constants
import { SIDEBAR_ITEMS } from '@/constants';

export const Sidebar = () => {
  const pathName = usePathname();

  return (
    <aside className='h-screen w-[195px] border-r-1 border-primary-200 border-opacity-50 px-[14px] pt-4'>
      <ul className='cursor-pointer'>
        {SIDEBAR_ITEMS?.map(({ href, label, Icon }) => {
          const liStyles =
            pathName === href ? 'mb-3 bg-foreground-300 bg-opacity-10' : '';

          return (
            <li
              key={label}
              className={`w-[167px] rounded-[6px] border-none py-2 pl-3 ${liStyles}`}
            >
              <Link
                href={href}
                className={`flex w-full items-center justify-start gap-3 border-none text-sm`}
              >
                <Icon />
                <span>{label}</span>
              </Link>
            </li>
          );
        })}
      </ul>
    </aside>
  );
};
