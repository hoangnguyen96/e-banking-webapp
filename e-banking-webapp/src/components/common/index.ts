export * from './Button';
export * from './Text';
export * from './Input';
export * from './RadioButton';
export * from './Navbar';
export * from './MenuDropdown';
export * from './Checkbox';
export * from './AnalyticsCard';
export * from './Select';
export * from './DailyLimit';
export * from './Calendar';
export * from './WizardForm';
export * from './Chip';
