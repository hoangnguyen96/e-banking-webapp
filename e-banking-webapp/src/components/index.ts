export * from './common';
export * from './icons';
export * from './Error';
export * from './NotFound';
export * from './Slideshow';
export * from './AuthContentWrapper';
export * from './LoginForm';
export * from './TargetTile';
export * from './QuickAction';
export * from './SignUpForm';
export * from './Charts';
export * from './Sidebar';
export * from './CreditCard';
export * from './Transfers';
export * from './ExpenseAnalysis';
export * from './StepProgress';
export * from './MyCards';
export * from './BalanceModal';
export * from './MyTarget';
export * from './ConnectedAccount';
export * from './SettingContainer/General';
export * from './DeleteAccount';
export * from './ServiceCard';
export * from './SettingContainer/Account';
export * from './DueTile';
export * from './MasterCard';
export * from './CardOverview';
export * from './MetricsCard';
export * from './Skeletons';
export * from './MyCalender';
