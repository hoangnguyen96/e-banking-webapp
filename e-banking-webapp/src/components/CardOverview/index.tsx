'use client';

// Libs
import { Card, CardBody } from '@nextui-org/react';

// Constants
import { createExpenseAnalysisOptions } from '@/constants';

// Mocks
import { MASTERCARD_CHART_MOCK } from '@/mocks';

// Component
import { Button, Text, CreditCard, MasterCard } from '@/components';

export const CardOverview = () => {
  return (
    <Card>
      <CardBody className='flex flex-row justify-between gap-[58px] p-0'>
        <div className='flex w-full flex-col gap-[14px] pl-4 pt-2'>
          <div className='flex items-center justify-between'>
            <Text as='span' className='text-sm font-semibold text-navyBlue'>
              My Cards
            </Text>
            <Button
              variant='outline'
              color='outline'
              radius='xs'
              className='h-auto w-auto border-navyBlue px-2 py-1 text-[10px] font-medium text-navyBlue'
              isDisabled
            >
              Add Card
            </Button>
          </div>
          <CreditCard
            cardNumber='537544114540'
            expireDate='06/24'
            holderName='DONALD FLINCH CORTEZ'
            bankName='Universal Bank'
          />
        </div>
        <MasterCard
          series={MASTERCARD_CHART_MOCK}
          totalBalance={createExpenseAnalysisOptions('$540,000')}
        />
      </CardBody>
    </Card>
  );
};
