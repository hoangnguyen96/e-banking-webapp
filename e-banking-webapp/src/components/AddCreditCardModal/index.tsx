'use client';

import { useEffect, useState } from 'react';
import Image from 'next/image';
import { z } from 'zod';
import { zodResolver } from '@hookform/resolvers/zod';
import { useForm } from 'react-hook-form';
import { AuthError, Session } from 'next-auth';

// Constants
import { ERROR_MESSAGES, IMAGES } from '@/constants';

// Interfaces
import { IAccount } from '@/interfaces';

// Actions
import { addNewCardByAccountId } from '@/actions';

// Schemas
import { CreditCardSchema } from '@/schemas';

// Services
import { getBalanceAccount } from '@/services';

// Context
import { useToastContext } from '@/context';

// Components
import * as WizardForm from '@/components/common/WizardForm';
import { Modal } from '../common';
import { AddCreditCard } from './AddCreditCard';
import { ConfirmAddCard } from './ConfirmAddCard';

interface IAddCreditCardModalProps {
  session: Session;
  isOpen: boolean;
  onClose: () => void;
}

type FormValues = z.infer<typeof CreditCardSchema>;

export const AddCreditCardModal = ({
  isOpen,
  session,
  onClose,
}: IAddCreditCardModalProps) => {
  const [accounts, setAccounts] = useState<IAccount[]>([]);
  const { showToast } = useToastContext();

  const form = useForm<FormValues>({
    resolver: zodResolver(CreditCardSchema),
    defaultValues: {
      fullName: '',
      cardNumber: '',
      expireAt: '',
      ccv: '',
      holderName: '',
    },
    reValidateMode: 'onBlur',
    mode: 'onBlur',
  });

  useEffect(() => {
    const fetchData = async () => {
      try {
        const { user } = await getBalanceAccount(String(session.user.id));

        setAccounts(user?.accounts || []);
      } catch (error) {
        if (error instanceof AuthError) {
          throw ERROR_MESSAGES.GET_ERROR;
        }
      }
    };

    fetchData();
  }, [session.user.id]);

  const submitHandler = async (accountId: string, data: FormValues) => {
    try {
      await addNewCardByAccountId(accountId, data);

      showToast(ERROR_MESSAGES.ADD_CARD_SUCCESS, 'success', 'top-center');

      return onClose();
    } catch (error) {
      showToast(ERROR_MESSAGES.ADD_CARD_FAILED, 'error', 'top-center');

      if (error instanceof AuthError) {
        throw ERROR_MESSAGES.ADD_CARD_FAILED;
      }
    }
  };

  return (
    <Modal
      isOpen={isOpen}
      onClose={onClose}
      size='4xl'
      placement='center'
      classNames={{
        base: 'p-0 rounded-xl bg-primary-200',
      }}
    >
      <div className='flex items-center justify-between'>
        <Image
          className='max-w-[50%] flex-1 px-[31px] py-[87px]'
          src={IMAGES.ADD_CREDIT_CARD_IMAGE}
          width={333}
          height={280}
          alt='Add Credit Card'
        />
        <div className='min-h-[500px] flex-1 rounded-xl bg-white px-4 py-8'>
          <WizardForm.Root
            schema={CreditCardSchema}
            form={form}
            className='flex grow flex-col'
          >
            <WizardForm.Step name='addCreditCard' key='addCreditCard'>
              <AddCreditCard />
            </WizardForm.Step>
            <WizardForm.Step name='addCreditCard' key='addCreditCard'>
              <ConfirmAddCard
                accounts={accounts}
                schema={CreditCardSchema}
                submitHandler={submitHandler}
              />
            </WizardForm.Step>
          </WizardForm.Root>
        </div>
      </div>
    </Modal>
  );
};
