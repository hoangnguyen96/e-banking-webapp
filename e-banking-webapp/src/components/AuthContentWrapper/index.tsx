'use client';

import { ReactNode } from 'react';
import { usePathname } from 'next/navigation';
import Link from 'next/link';
import Image from 'next/image';

// Constants
import { IMAGES, ROUTES } from '@/constants';

// Interfaces
import { TEXT_SIZE, TEXT_VARIANT } from '@/interfaces';

// Components
import { Text } from '../common';

interface AuthContentWrapperProps {
  textFooter?: string;
  textHeading: ReactNode;
  formContent: ReactNode;
}

export const AuthContentWrapper = ({
  textHeading,
  formContent,
  textFooter,
}: AuthContentWrapperProps) => {
  const pathname = usePathname();

  const isOnSignIn = pathname === ROUTES.SIGN_IN;

  const linkDetails = {
    text: isOnSignIn ? 'Register' : 'Sign In',
    href: isOnSignIn ? ROUTES.SIGN_UP : ROUTES.SIGN_IN,
  };

  const classLogo = isOnSignIn ? 'mt-0 md:mt-[108px]' : '';

  return (
    <div className='mx-auto flex w-full max-w-lg flex-1 flex-col'>
      {/* Logo */}
      <Image
        className={`pb-9 ${classLogo}`}
        src={IMAGES.LOGO}
        width={104}
        height={36}
        alt='Logo EBanking'
      />

      <div className='mb-7 max-w-[406px]'>
        <Text as='h1' size={TEXT_SIZE['4XL']} variant={TEXT_VARIANT.DEFAULT}>
          {textHeading}
        </Text>
      </div>

      {/* Form */}
      <div className='mb-[17px] max-w-[325px]'>{formContent}</div>

      {textFooter && (
        <Text
          size={TEXT_SIZE.SM}
          variant={TEXT_VARIANT.DEFAULT}
          className='leading-[20px]'
        >
          {textFooter}&nbsp;
          <Link
            href={linkDetails.href}
            className='leading-[20px] text-primary-200'
          >
            {linkDetails.text}
          </Link>
        </Text>
      )}
    </div>
  );
};
