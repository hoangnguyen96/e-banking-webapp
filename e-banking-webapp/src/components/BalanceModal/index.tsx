'use client';

import { CircularProgress } from '@nextui-org/react';

// Constants
import { createExpenseAnalysisOptions } from '@/constants';

// Interfaces
import { TEXT_SIZE, TEXT_VARIANT } from '@/interfaces';

// Mocks
import { MOCK_SERIES_EXPENSE_ANALYSIS } from '@/mocks';

// Components
import { Modal, Text } from '../common';
import { MyCards } from '../MyCards';
import { ExpenseAnalysis } from '../ExpenseAnalysis';

interface BalanceModalProps {
  username?: string;
  currentBalance?: string;
  totalInvestment?: string;
  totalBalance?: string;
  totalIncome?: string;
  percent?: number;
  isOpen?: boolean;
  onClose: () => void;
}

const BalanceModal = ({
  username = '',
  currentBalance = '',
  totalInvestment = '',
  totalBalance = '',
  totalIncome = '',
  percent = 0,
  isOpen = false,
  onClose,
}: BalanceModalProps) => {
  return (
    <Modal
      isOpen={isOpen}
      onClose={onClose}
      size='4xl'
      placement='center'
      classNames={{ base: 'p-0 rounded-xl' }}
    >
      <div className='absolute left-4 top-4'>
        <Text as='h4' className='text-base font-medium text-primary-200'>
          Good Evening,{' '}
          <Text
            as='span'
            variant={TEXT_VARIANT.DEFAULT}
            className='font-normal'
          >
            {username}
          </Text>
        </Text>
      </div>
      <div className='flex rounded-xl bg-lightGraySolid'>
        <div className='flex flex-1 flex-col rounded-xl bg-white pb-[48px] pl-3 pr-[23px] pt-[56px]'>
          <div className='flex items-center justify-between'>
            <div className='flex flex-col gap-2'>
              <Text
                as='h5'
                className='text-sm font-semibold'
                variant={TEXT_VARIANT.DEFAULT}
              >
                Current Balance
              </Text>
              <Text
                as='span'
                className='font-extrabold'
                variant={TEXT_VARIANT.DEFAULT}
              >
                ${currentBalance}
                <Text
                  as='span'
                  size={TEXT_SIZE['XS']}
                  className='font-semibold text-black/50'
                >
                  /${totalIncome}
                </Text>
              </Text>
            </div>
            <div className='flex flex-col gap-2'>
              <Text
                as='h5'
                className='text-sm font-semibold'
                variant={TEXT_VARIANT.DEFAULT}
              >
                Total Investments
              </Text>
              <Text
                as='span'
                className='font-extrabold'
                variant={TEXT_VARIANT.DEFAULT}
              >
                ${totalInvestment}
              </Text>
            </div>
          </div>

          <div className='relative mx-auto mt-9 flex h-[200px] w-[200px] items-center justify-center'>
            <CircularProgress
              classNames={{
                svg: 'w-[200px] h-[200px] drop-shadow-md',
                indicator: 'stroke-[#2A9D8F]',
                track: 'stroke-[#2A9D8F]/10',
                value: 'text-medium font-extrabold text-black',
              }}
              showValueLabel={false}
              strokeWidth={3}
              value={percent}
            />
            <div className='absolute flex flex-col items-center'>
              <Text
                className='text-center text-sm font-medium'
                variant={TEXT_VARIANT.DEFAULT}
              >
                Spend
              </Text>
              <Text
                as='span'
                className='text-center font-extrabold'
                variant={TEXT_VARIANT.DEFAULT}
              >
                {percent}%
              </Text>
            </div>
          </div>
        </div>

        {/* My Cards and Expense Analysis */}
        <div className='flex flex-1 flex-col pb-[28px] pl-4 pr-3 pt-[56px]'>
          <Text as='h4' variant={TEXT_VARIANT.DEFAULT}>
            My Cards
          </Text>
          <div className='my-4'>
            <MyCards />
          </div>
          <Text
            size={TEXT_SIZE['XS']}
            variant={TEXT_VARIANT.DEFAULT}
            className='max-w-[180px] font-normal'
          >
            You have a total of{' '}
            <Text
              as='span'
              size={TEXT_SIZE['XS']}
              variant={TEXT_VARIANT.DEFAULT}
              className='font-bold'
            >
              ${totalBalance}
            </Text>{' '}
            available in your wallet
          </Text>

          <div className='mt-2.5'>
            <ExpenseAnalysis
              options={createExpenseAnalysisOptions('220 000')}
              series={MOCK_SERIES_EXPENSE_ANALYSIS}
            />
          </div>
        </div>
      </div>
    </Modal>
  );
};

export default BalanceModal;
