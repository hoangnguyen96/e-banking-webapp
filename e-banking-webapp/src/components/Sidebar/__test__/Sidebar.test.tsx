// Libs
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { usePathname } from 'next/navigation';

// Components
import { Sidebar } from '@/components';

jest.mock('next/navigation', () => ({
  usePathname: jest.fn(),
}));

const mockSidebarItems = [
  { label: 'General', href: '/setting/general', icon: <span>Icon1</span> },
  { label: 'Account', href: '/setting/account', icon: <span>Icon2</span> },
];

describe('Sidebar component', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  test('renders correctly with sidebar items', () => {
    (usePathname as jest.Mock).mockReturnValue('/setting/general');

    const { container } = render(<Sidebar sidebarItem={mockSidebarItems} />);

    expect(container).toMatchSnapshot();
  });

  test('renders all sidebar items', () => {
    (usePathname as jest.Mock).mockReturnValue('/');

    render(<Sidebar sidebarItem={mockSidebarItems} />);

    expect(screen.getByText('General')).toBeInTheDocument();
    expect(screen.getByText('Account')).toBeInTheDocument();
  });
});
