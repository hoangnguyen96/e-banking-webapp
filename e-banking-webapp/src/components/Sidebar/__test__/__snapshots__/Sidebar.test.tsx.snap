// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar component renders correctly with sidebar items 1`] = `
<div>
  <aside
    class="h-screen w-[195px] border-r-1 border-semiTransparentDimGray px-[14px] pt-6"
  >
    <a
      href="/setting/general"
    >
      <button
        class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-6 min-w-24 h-12 gap-3 [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none data-[hover=true]:opacity-hover border-transparent bg-overlay rounded-md w-full max-h-10 py-2.5 text-sm text-black"
        type="button"
      >
        <span
          aria-hidden="true"
          focusable="false"
          tabindex="-1"
        >
          Icon1
        </span>
        General
      </button>
    </a>
    <a
      href="/setting/account"
    >
      <button
        class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-6 min-w-24 h-12 gap-3 [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none data-[hover=true]:opacity-hover border-transparent bg-none rounded-md w-full max-h-10 py-2.5 text-sm text-black"
        type="button"
      >
        <span
          aria-hidden="true"
          focusable="false"
          tabindex="-1"
        >
          Icon2
        </span>
        Account
      </button>
    </a>
  </aside>
</div>
`;
